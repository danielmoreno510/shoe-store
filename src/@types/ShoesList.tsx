type ShoesListProps = {
  isFetchingShoesList: boolean;
  shoesList: Shoe[];
  cart: Shoe[];
  getShoes: any;
  setAddToCart(shoe: Shoe): any;
  setCurrentShoe(shoe: Shoe): any;
  navigation: Navigate;
};
