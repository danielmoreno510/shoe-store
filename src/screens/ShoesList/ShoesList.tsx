import React from 'react';
import {View, FlatList} from 'react-native';
import {Spinner} from 'native-base';
import {isEmpty} from 'lodash';

import Shoe from '../../components/common/Shoe/Shoe';
import styles from './styles';
import {primaryColor} from '../../constants';

const ShoesList: React.FC<ShoesListProps> = ({
  isFetchingShoesList,
  shoesList,
  cart,
  getShoes,
  setAddToCart,
  setCurrentShoe,
  navigation,
}) => {
  React.useEffect(() => {
    isEmpty(shoesList) && getShoes();
  }, [shoesList, getShoes]);

  const addToCart = (shoe: Shoe) => setAddToCart(shoe);
  const showProduct = (shoe: Shoe) => {
    navigation.navigate('Shoe Detail');
    setCurrentShoe(shoe);
  };

  return (
    <View style={styles.container}>
      {!isFetchingShoesList ? (
        <FlatList
          data={shoesList}
          renderItem={({item}) => (
            <Shoe shoe={item} cart={cart} addToCart={addToCart} showProduct={showProduct} />
          )}
          keyExtractor={item => item.id}
          showsVerticalScrollIndicator={false}
        />
      ) : (
        <Spinner color={primaryColor} />
      )}
    </View>
  );
};

export default ShoesList;
