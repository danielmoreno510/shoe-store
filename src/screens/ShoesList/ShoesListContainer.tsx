import ShoesList from './ShoesList';
import {StoreService} from '../../redux/StoreService';
import {fetchShoesList, putAddToCart, putCurrentShoe} from '../../redux/actions';

export const mapStateToProps = state => ({
  isFetchingShoesList: state.shoes.isFetchingShoesList,
  shoesList: state.shoes.shoesList,
  cart: state.shoes.cart,
});

export const mapDispatchToProps = dispatch => ({
  getShoes: () => {
    dispatch(fetchShoesList());
  },
  setAddToCart: shoe => {
    dispatch(putAddToCart(shoe));
  },
  setCurrentShoe: shoe => {
    dispatch(putCurrentShoe(shoe));
  },
});

const ShoesListContainer = StoreService.connect(mapStateToProps, mapDispatchToProps)(ShoesList);

export default ShoesListContainer;
