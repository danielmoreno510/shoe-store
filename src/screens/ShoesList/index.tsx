import ShoesListContainer from './ShoesListContainer';

export default ShoesListContainer;
