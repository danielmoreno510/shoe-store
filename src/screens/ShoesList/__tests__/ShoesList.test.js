import React from 'react';
import {shallow} from 'enzyme';

import ShoesList from '../ShoesList';

jest.spyOn(React, 'useEffect').mockImplementation(f => f());

describe('ShoesList component', () => {
  describe('GIVEN a set of valid props', () => {
    const testProps = {
      isFetchingShoesList: true,
      shoesList: [],
      cart: [{}],
      getShoes: jest.fn(),
      setAddToCart: jest.fn(),
      setCurrentShoe: jest.fn(),
      navigation: {navigate: jest.fn()},
    };

    describe('WHEN the component is rendered', () => {
      const shallowWrapper = shallow(<ShoesList {...testProps} />);

      it('THEN should display a regular ShoesList', () => {
        expect(shallowWrapper).toMatchSnapshot();
      });
    });

    describe('WHEN the addToCart and showProduct functions are called', () => {
      const customProps = {...testProps, isFetchingShoesList: false, shoesList: [{}]};
      const shallowWrapper = shallow(<ShoesList {...customProps} />);
      const flatList = shallowWrapper.find('FlatList');
      const item = flatList.renderProp('renderItem')({});
      item.props().addToCart();
      item.props().showProduct();
      flatList.props().keyExtractor({item: {id: 1}});

      it('THEN setAddToCart and setCurrentShoe function should be called', () => {
        expect(testProps.setAddToCart).toBeCalledTimes(1);
        expect(testProps.setCurrentShoe).toBeCalledTimes(1);
      });
    });
  });
});
