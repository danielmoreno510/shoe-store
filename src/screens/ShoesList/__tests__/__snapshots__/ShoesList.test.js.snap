// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoesList component GIVEN a set of valid props WHEN the component is rendered THEN should display a regular ShoesList 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 37.5,
    }
  }
>
  <Component
    color="#f1b30a"
  />
</View>
`;
